src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -764,31 +767,56 @@ HNSWIndex<DataType, DistType>::searchLayer(idType ep_id, const void *data_point,
     return top_candidates;
 }
 
+template <typename DataType, typename DistType>
+idType
+HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(candidatesList<DistType> &top_candidates,
+                                                        const size_t M) const {
+    idType best;
+    if (top_candidates.size() < M) {
+        best = std::min_element(top_candidates.begin(), top_candidates.end(),
+                                [](const auto &a, const auto &b) { return a.first < b.first; })
+                   ->second;
+    } else {
+        getNeighborsByHeuristic2_internal<false>(top_candidates, M, best, nullptr);
+    }
+    return best;
+}
+
 template <typename DataType, typename DistType>
 void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
-    candidatesMaxHeap<DistType> &top_candidates, const size_t M) {
+    candidatesList<DistType> &top_candidates, const size_t M,
+    vecsim_stl::vector<idType> &not_chosen) const {
+    idType dummy;
+    getNeighborsByHeuristic2_internal<true>(top_candidates, M, dummy, &not_chosen);
+}
+
+template <typename DataType, typename DistType>
+template <bool record_not_chosen>
+void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2_internal(
+    candidatesList<DistType> &top_candidates, const size_t M, idType &best,
+    vecsim_stl::vector<idType> *not_chosen) const {
```

Comment:
```suggestion
    vecsim_stl::vector<idType> *to_remove) const {
```

Reply:
this list may contain candidates that were not neighbors in the first place.
How about `removed_candidates`? in the context of this function it makes some sense